google/cloud/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -0,0 +1,275 @@
+      impl_->Finish().then([this](future<Status> finish_future) {
+        Status finish_status = finish_future.get();
+        if (finish_status.ok()) {
+          this->finish_status_ = absl::make_optional(finish_status);
+          return make_ready_future(true);
+        }
+        auto const retry_policy = this->retry_policy_prototype_->clone();
+        auto const backoff_policy = this->backoff_policy_prototype_->clone();
+        return this->RetryLoop(retry_policy, backoff_policy, RetryType.Write);
+      });
+    });
+  }
+
+  future<Status> Finish() override {
+    std::lock_guard<std::mutex> g{mu_};
+    if (finish_status_.has_value()) {
+      return make_ready_future(finish_status_.value());
+    }
+    return impl_->Finish();
+  }
+
+  future<Status> Status() {
+    std::lock_guard<std::mutex> g{mu_};
+    return make_ready_future(current_status_);
+  }
+
+ private:
+  enum class RetryType { Status, Read, Write };
+
+  absl::variant<future<absl::optional<ResponseType>>, future<bool>> RetryLoop(
+      std::shared_ptr<RetryPolicy const> retry_policy,
+      std::shared_ptr<BackoffPolicy const> backoff_policy,
+      RetryType retry_type) {
+    std::lock_guard<std::mutex> g{mu_};
+    if (!retry_policy->IsExhausted() && !is_canceled) {
+      return this->sleeper_(backoff_policy->OnCompletion())
+          .sleep()
+          .then([this, retry_policy, backoff_policy](future<void>) {
+            std::lock_guard<std::mutex> g{this->mu_};
+            this->impl_ = this->stream_factory_();
+            return this->impl_->Start().then([this](future<bool> start_future) {
+              if (!start_future.get()) {
+                return this->RetryLoop(retry_policy, backoff_policy,
+                                       retry_type);
+              }
+              if (retry_type == RetryType.Start) {
+                return make_ready_future(true);
+              }
+              return this->reinitializer_(std::move(this->impl_))
+                  .then([this, retry_policy, backoff_policy](
+                            future<StatusOr<
+                                std::unique_ptr<AsyncStreamingReadWriteRpc<
+                                    RequestType, ResponseType>>>>
+                                reinitialize_future) {
+                    StatusOr<std::unique_ptr<
+                        AsyncStreamingReadWriteRpc<RequestType, ResponseType>>>
+                        reinitialize_response = reinitialize_future.get();
+                    if (!reinitialize_response.ok()) {
+                      return this->RetryLoop(retry_policy, backoff_policy,
+                                             retry_type);
+                    }
+                    std::lock_guard<std::mutex> g{this->mu_};
+                    this->impl_ = std::move(*reinitialize_response);
+                    this->current_status_ = Status(
+                        StatusCode.kDataLoss, "Stream failed, May try again");
```

Comment:
never use DataLoss code. It makes users think we lost their messages in the server :(